firedrake/slate/slac/compiler.py

```diff
@@ -342,15 +350,25 @@ def tensor_assembly_calls(builder):
         # (exterior) facets will be present within the loop. The
         # cell facets are labelled `1` for interior, and `0` for
         # exterior.
-
         statements.append(ast.FlatBlock("/* Loop over cell facets */\n"))
-        int_calls = list(chain(*[builder.assembly_calls[it_type]
+        int_calls = list(chain(*[assembly_calls[it_type]
                                  for it_type in ("interior_facet",
                                                  "interior_facet_vert")]))
-        ext_calls = list(chain(*[builder.assembly_calls[it_type]
+        ext_calls = list(chain(*[assembly_calls[it_type]
                                  for it_type in ("exterior_facet",
                                                  "exterior_facet_vert")]))
 
+        # Generate logical statements for handling *exterior facet*
+        # integrals on subdomains (currently this is the only supported type)
+        if assembly_calls["subdomains_exterior_facet"]:
+            subdomain_calls = assembly_calls["subdomains_exterior_facet"]
+            for k in subdomain_calls:
```

Comment:
Why not just `for k in assembly_calls["subdomains_exterior_facet"]` ?

Reply:
Oh, I see, because you index later on.  I think a clearer way to write this would be to collect a list of 2-tuples of `(subdomain, integral_call)` and then do:

```python
from gem.utils import groupby

for (subdomain, calls) in groupby(assembly_calls["subdomains_exterior_facet"], lambda x: x[0]):
    ext_calls.append(ast.If(ast.Eq(ast.Symbol(..., subdomain)), (as.Block(calls, ...), )))))
```

Reply:
>I think a clearer way to write this would be to collect a list of 2-tuples of `(subdomain, integral_call)` and then do:
>
>```python
>from gem.utils import groupby
>
>for (subdomain, calls) in groupby(assembly_calls["subdomains_exterior_facet"], lambda x: x[0]):
>     ext_calls.append(ast.If(ast.Eq(ast.Symbol(..., subdomain)), (as.Block(calls, ...), )))))
>```

Ah, I like this.